system/lib/wasmfs/wasmfs.h

```diff
@@ -0,0 +1,55 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the global state of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include "file_table.h"
+#include <assert.h>
+#include <emscripten/html5.h>
+#include <map>
+#include <mutex>
+#include <sys/stat.h>
+#include <vector>
+#include <wasi/api.h>
+
+namespace wasmfs {
+
+class WasmFS {
+  // Files will be preloaded in this constructor.
+  WasmFS() {}
+
+  // Since the constructor is private, the only way to access the global file
+  // state is through this get method.
+  static WasmFS& get() {
+    static WasmFS globalState;
+    return globalState;
+  }
```

Comment:
I was thinking we don't need this. Rather, at the top level, we could have
```cpp
class WasmFS {
  ...
};

WasmFS wasmFS;
```
So, no statics, no special `get()` method, just a single global singleton. The compiler will initialize it for us at the proper time.

(A possible issue might be if we have any special ctors that depend on files, we might need to add a priority attribute here.)

Reply:
1. Is there any concern about refactoring in the future - i.e. having to change the priority order?
2. This approach doesn't have lazy initialization on the global state constructor. I guess we could have lazy initialization of the `FileTable` and `RootDirectory` members however using `std::optional` which Thomas brought up.

Reply:
Btw, note that a nice thing with using priorities here is that it lets the WasmFS class itself not worry about the order of init at all: the attribute is on the instance, and not the class. If in the future we have more WasmFS instances, they might not need such an attribute, etc. That is, when one creates an instance one needs to think about this, and not at the class level, which is more modular (each instance can do the right thing for it).

Reply:
I wonder if lazy initialization has any benefit for DCE. But absent any data to the contrary, removing lazy initialization and optimizing for code simplicity seems fine to me.

Reply:
This approach works with the below:

```
# 19 "wasmfs.cpp" 3
__attribute__((init_priority(100))) WasmFS wasmFS;
# 21 "wasmfs.cpp"
```

I added a large comment about the usage here. The only downside is that `wasmFS` can be the only static global object defined. I had to remove the static `entries` from `file_table.cpp` as the initialization order was undefined. The upside is that we should not require any of these objects.